kernel/src/hil/i2c.rs

```diff
@@ -53,6 +58,73 @@ pub trait I2CMaster {
     fn write_read(&self, addr: u8, data: &'static mut [u8], write_len: u8, read_len: u8);
     fn write(&self, addr: u8, data: &'static mut [u8], len: u8);
     fn read(&self, addr: u8, buffer: &'static mut [u8], len: u8);
+
+    /// Write data then read data via the I2C Master device in an SMBus
+    /// compatible way.
+    ///
+    /// This function will use the I2C master to write data to a device and
+    /// then read data from the device in a SMBus compatible way. This will be
+    /// a best effort attempt to match the SMBus specification based on what
+    /// the hardware can support.
+    /// This function is expected to make any hardware changes required to
+    /// support SMBus and then revert those changes to support future I2C.
+    ///
+    /// addr: The address of the device to write to
+    /// data: The buffer to write the data from and read back to
+    /// write_len: The length of the write operation
+    /// read_len: The length of the read operation
+    #[allow(unused_variables)]
+    fn smbus_write_read(
```

Comment:
I'm a little conflicted on these and I would very much appreciate thoughts from others.

It doesn't seem to me that all I2C masters should be required to implement SMBus functionality. If that's the case, then we should probably split the SMBus stuff into a separate trait, where an I2C master could implement either both or only I2CMaster.

However, you already provided defaults for the SMBus functions, which means it is allowable to not implement them. Further, by modifying the virtual_i2c to require SMBus support (even if just to return an error), you're essentially requiring that all I2C implementations would have to support both. In practice with could be a single `impl SMBusMaster for XXX {}` (which would use the default implementations) for everything, which seems pointless.

Also, I don't think it was a bad idea to include SMBus devices as part of the virtual_i2c. You could imagine a copy of the code that was `virtual_i2c_smbus` and requires an i2c that implements the SMBus trait, but it would just be code duplication that has upkeep problems.

So I'm left unsure how I feel.

Reply:
I also felt conflicted about splitting them. If split then it makes the `virtual_i2c` support much more difficult. Currently we can mux `I2CDevices` and `SMBusDevices` with the `I2CMaster` driver. If we split the implementations (but the underlying hardware will be a single device) then we can't mux virtual_i2c and virtual_smbus without another layer of virtulisation. So it becomes much more complex.

Splitting the two would required more virtulisation without much gain, as I don't see any chips that support just a SMBus device.

Reply:
Well, you could split them and then `virtual_i2c` could require an i2c that implements `I2CMaster + SMBusMaster`. That would allow them to be split without requiring any additional verification. But that's also the case above where every i2c chip driver ends up with the additional line: `impl SMBusMaster for XXX {};` just to implement the default functions for SMBus that return errors so that they work with `virtual_i2c`, which seems pretty pointless.

Reply:
This implementation solves my concerns and it avoids too much duplication!